File: .github/scripts/vectorize_recipes.py
import json
import os
from pathlib import Path
from sentence_transformers import SentenceTransformer
import numpy as np

# Configuración
MODEL_NAME = 'all-MiniLM-L6-v2'  # Cambia por otro modelo si lo prefieres
INPUT_JSON = 'recipes_metadata.json'
OUTPUT_JSONL = 'recipes_vectors.jsonl'

# Campos a vectorizar (puedes ajustar según tu estructura)
FIELDS_TO_VECTORIZE = [
    'nombre', 'descripcion', 'ingredientes', 'preparacion', 'analisis', 'sabiduria_colectiva', 'fuentes'
]

def load_metadata(json_path):
    with open(json_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def get_text_for_vectorization(metadata):
    texts = []
    for field in FIELDS_TO_VECTORIZE:
        value = metadata.get(field)
        if value:
            if isinstance(value, list):
                texts.extend([str(v) for v in value])
            else:
                texts.append(str(value))
    return '\n'.join(texts)

def main():
    if not os.path.exists(INPUT_JSON):
        print(f"No se encontró {INPUT_JSON}. Ejecuta primero el script de extracción de metadatos.")
        return
    data = load_metadata(INPUT_JSON)
    model = SentenceTransformer(MODEL_NAME)
    with open(OUTPUT_JSONL, 'w', encoding='utf-8') as fout:
        for recipe in data:
            text = get_text_for_vectorization(recipe)
            embedding = model.encode(text).tolist()
            out = {
                'id': recipe.get('file', ''),
                'metadata': recipe,
                'embedding': embedding
            }
            fout.write(json.dumps(out, ensure_ascii=False) + '\n')
    print(f"Vectorización completada. Archivo generado: {OUTPUT_JSONL}")

if __name__ == "__main__":
    main()

File: .github/scripts_ts/vectorize_recipes.ts
import { pipeline } from '@xenova/transformers';
import fs from 'fs/promises';
import path from 'path';

// Configuración
const MODEL_NAME = 'Xenova/all-MiniLM-L6-v2'; // Modelo equivalente a 'all-MiniLM-L6-v2'
const INPUT_JSON = path.join(__dirname, '../../recipes_metadata.json');
const OUTPUT_JSONL = path.join(__dirname, '../../recipes_vectors.jsonl');

// Campos a vectorizar (ajusta según tu estructura)
const FIELDS_TO_VECTORIZE = [
  'nombre', 'descripcion', 'ingredientes', 'preparacion', 'analisis', 'sabiduria_colectiva', 'fuentes'
];

interface RecipeMetadata {
  [key: string]: any;
}

async function loadMetadata(jsonPath: string): Promise<RecipeMetadata[]> {
  const content = await fs.readFile(jsonPath, 'utf8');
  return JSON.parse(content);
}

function getTextForVectorization(metadata: RecipeMetadata): string {
  const texts: string[] = [];
  for (const field of FIELDS_TO_VECTORIZE) {
    const value = metadata[field];
    if (value) {
      if (Array.isArray(value)) {
        texts.push(value.join('\n'));
      } else {
        texts.push(String(value));
      }
    }
  }
  return texts.join('\n');
}

async function main() {
  try {
    // Verificar si el archivo de entrada existe
    await fs.access(INPUT_JSON);

    // Cargar metadatos
    const data = await loadMetadata(INPUT_JSON);

    // Inicializar el modelo de embeddings
    const extractor = await pipeline('feature-extraction', MODEL_NAME);

    // Abrir el archivo de salida
    const fout = await fs.open(OUTPUT_JSONL, 'w');

    // Procesar cada receta
    for (const recipe of data) {
      const text = getTextForVectorization(recipe);
      // Generar embeddings
      const output = await extractor(text, { pooling: 'mean', normalize: true });
      const embedding = Array.from(output.data);

      // Escribir en el archivo JSONL
      await fout.appendFile(JSON.stringify({
        ...recipe,
        vector: embedding
      }) + '\n');
    }

    await fout.close();
    console.log('Vectorización completada.');
  } catch (err) {
    console.error('Error:', err);
  }
}

main();
